### PR TITLE
Add ability to import CAR files and retrieve to a CAR file

### PR DESCRIPTION
**Pull request by hannahhoward**
# Goals

Support easy file transfer to IPFS by importing/exporting CAR files

# Implementation

- Change the ClientImport and ClientRetrieve commands to take a FileRef, which specifies both Path and whether to treat the file as a CAR
- Add a "car" boolean option to the `client import` and `client retrieve` commands at the command line

### CODE REVIEW

File: api/api_full.go [hunk 1]
```diff
@@ -153,6 +153,11 @@ type FullNode interface {
 	PaychVoucherSubmit(context.Context, address.Address, *paych.SignedVoucher) (cid.Cid, error)
 }
 
+type FileRef struct {
+	Path  string
+	IsCAR bool
```
Comment (whyrusleeping) [inline api/api_full.go]: Let's take this one step further and just have a stringly typed 'ImportType' field, which we can set to car. I imagine there will be other potentials here in the future

File: node/impl/client/client.go [hunk 2]
```diff
@@ -196,17 +197,31 @@ func (a *API) ClientFindData(ctx context.Context, root cid.Cid) ([]api.QueryOffe
...
-func (a *API) ClientImport(ctx context.Context, path string) (cid.Cid, error) {
-	f, err := os.Open(path)
+func (a *API) ClientImport(ctx context.Context, ref api.FileRef) (cid.Cid, error) {
+	f, err := os.Open(ref.Path)
 	if err != nil {
 		return cid.Undef, err
 	}
+
+	if ref.IsCAR {
+		result, err := car.LoadCar(a.Blockstore, f)
```
Comment (whyrusleeping) [inline node/impl/client/client.go]: Just a note, but this will meant that importing a car file doesnt use the filestore

Comment (hannahhoward) [reply node/impl/client/client.go]: yea I saw that.

**Comment by hannahhoward** [comment]
I believe we need to make changes to `go-car` to support importing by reference -- it does not currently have any ability to work with a filestore

**Comment by hannahhoward** [comment]
@whyrusleeping this now uses the filestore instead of importing and I went ahead and added a PR to go-car to support writing filestore refs properly.

Note: I had to add some hacking to ClientListImports to get the car CIDs to show up -- because the first block in a car file is not at offset 0 -- this is purely for cosmetic reasons so the CAR file shows when you do `lotus client local` -- and if this is not super important it can be removed.

**Pull request merged**